### PR TITLE
DC-2431/DC-2447 encrypted return url and return link text support

### PR DESCRIPTION
**Pull request by elabeca**

### CODE REVIEW

File: app/controllers/EnquiryController.scala [hunk 1]
```diff
@@ -27,6 +27,8 @@ import play.api.mvc.{ Action, AnyContent, Request, Result }
 import uk.gov.hmrc.auth.core.AuthProvider.GovernmentGateway
 import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
 import uk.gov.hmrc.auth.core.{ AuthConnector, AuthProviders, AuthorisationException }
+//import uk.gov.hmrc.crypto.Crypted
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:30]: remove commented line ?

File: app/controllers/EnquiryController.scala [hunk 2]
```diff
@@ -50,7 +53,10 @@ class EnquiryController @Inject()(
 
   val form: Form[EnquiryDetails] = formProvider()
 
-  def onPageLoad(enquiryType: String): Action[AnyContent] =
+  def onPageLoad(
+    enquiryType: String,
+    encryptedReturnUrl: Option[String],
+    encodedReturnUrlText: Option[String]): Action[AnyContent] =
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:59]: i think these two options should be just one, maybe better if we combine them into a case class:
ReturnUrl(encryptedLink: String, encodedText: String).
Then validate that both are present or both missing if otherwise return BadRequest.

Comment (elabeca) [reply app/controllers/EnquiryController.scala:59]: Could we update the story to record these new requirements and I'll update.

Comment (elabeca) [reply app/controllers/EnquiryController.scala:59]: I've handled the above (described in the it tests) but by providing a default label if not provided. I can definately return a BadRequest.

Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:58]: update the requirements on the ticket.

**Comment by elabeca** [comment]
Please do not merge yet - possible requirement changes on Monday morning after 9:30 meeting

File: it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala [hunk 3]
```diff
@@ -159,6 +160,100 @@ class EnquiryControllerFrontendSpec
...
+      textArea("question").value = "A question from the customer"
+      click on find(id("submit")).value
+      eventually {
+        pageSource must include(
+          "<a href=\"/two-way-message-frontend/confirmation-redirect?enquiryType=sa-general&amp;returnUrl=%2Fsome-return%2Furl&amp;returnLinkText=Q2xpY2sgaGVyZSB0byBiZSByZWRpcmVjdGVk\">")
+        pageSource must include("Click here to be redirected")
+      }
+    }
+
+    "Attempting to send a valid SA-GENERAL related message with a return url but no return link text query parameters should return a BAD REQUEST" in {
```
Comment (elabeca) [inline it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala:210]: need to fix this test

File: it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala [hunk 4]
```diff
@@ -159,6 +160,135 @@ class EnquiryControllerFrontendSpec
...
+      go to s"http://localhost:$port/two-way-message-frontend/message/sa-general/make_enquiry?returnUrl=/some-return/url&returnLinkText=Q2xpY2sgaGVyZSB0byBiZSByZWRpcmVjdGVk"
+      textField("subject").value = "A subject"
+      textField("telephone").value = "07700 900077"
+      emailField("email").value = "test@dummy.com"
+      textArea("question").value = "A question from the customer"
+      click on find(id("submit")).value
+      eventually {
+        pageSource must include(
+          "<a id=\"redirect-return-link\" href=\"/some-return/url\" target=\"_self\" data-sso=\"false\">")
+        pageSource must include("Click here to be redirected")
```
Comment (elabeca) [inline it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala:206]: - [ ] handle encryption of returnLinktText

File: conf/messages [hunk 5]
```diff
@@ -60,12 +61,11 @@ enquiry.question.hint = More detail will help us answer your question quickly.
...
 enquiry.success.next.heading = What happens next
-enquiry.success.next.para = We are working on your reply.
-enquiry.success.next.para1 = You do not need to follow up by sending another message or by calling us.
-enquiry.success.next.para2 = When your reply is ready, we will send you an email to let you know.
-enquiry.success.next.para3 = You can
-enquiry.success.next.para4 = by signing in to HMRC services.
+enquiry.success.next.para1 = We will aim to reply to you within
+enquiry.success.next.para1_1 = using the contact details you provided.
+enquiry.success.next.para2 = Because we are managing a lot of requests, do not contact HMRC by phone about this question, or send another message/
 enquiry.success.next.link_text = view your messages
```
Comment (rwalpole) [inline conf/messages:69]: I think you may need to capitalise this, i.e. `View your messages` as previously it was being used inside a sentence and now it appears on it's own.

File: app/views/enquiry_submitted.scala.html [hunk 6]
```diff
@@ -27,13 +29,17 @@
...
     </div>
+    @components.text("enquiry.success.note.para1")
     <h2 class="heading-medium">@messages("enquiry.success.next.heading")</h2>
-    @components.text("enquiry.success.next.para")
-    @components.text("enquiry.success.next.para1")
-    @components.text("enquiry.success.next.para2")
-    <p>@messages("enquiry.success.next.para3") <a href="@routes.EnquiryController.messagesRedirect()">
-       @messages("enquiry.success.next.link_text")</a> @messages("enquiry.success.next.para4")</p>
-    <a id="start-again" href="@routes.EnquiryController.personalAccountRedirect()" class="button">@messages("site.account.home")</a>
+    <p>@components.text("enquiry.success.next.para1") @responseTime @components.text("enquiry.success.next.para1_1")</p>
```
[the remark this answers is not in the document]
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:34]: "We will aim to reply to you within X days using the contact details you provided" seems to be in its own P tag on https://jira.tools.tax.service.gov.uk/browse/DC-2447 screenshot

File: app/views/enquiry_submitted.scala.html [hunk 7]
```diff
@@ -27,13 +29,17 @@
...
+    <p>@components.text("enquiry.success.next.para2")</p>
+    <p>@{
+        returnLink match {
+            case Some(link)                             => Link.toInternalPage(id = Some("redirect-return-link"), url = link.returnUrl, value = Some(link.returnLinkText)).toHtml
+            case _ if (enquiryType.startsWith("p800"))  => Link.toInternalPage(id = Some("redirect-return-link"), url = appConfig.ptaInboxUrl, value = Some(messages("enquiry.success.next.link_text"))).toHtml
```
[the remark this answers is not in the document]
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:39]: That's correct. There are tests covering the "NOT_FOUND" for P800 that I've added earlier today, but following the discussion we had yesterday morning, I was asked to make sure that should we re-enable the route, it follows the story requirement of "sending BTA users to default /personal-account/messages if return url is not provided" (albeit having the disabled path in our routes)

File: conf/messages (hunk 5 above)
Comment (elabeca) [reply conf/messages]: Actually it's the incorrect text altogether, thanks for highlighting this.

File: app/views/enquiry_submitted.scala.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (rwalpole) [reply app/views/enquiry_submitted.scala.html:34]: Yes but this is how it was before and what I am saying is you seem to have added extra `<p>` tags so that "X days" will appear on it's own line.

[the remark this answers is not in the document]
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:34]:
```
<p>@components.text("enquiry.success.next.para1") @responseTime @components.text("enquiry.success.next.para1_1")</p>
```
I'm not sure I follow. First we have the P tag, then we have "We will aim to reply to you within " then we have "X days" then we have " using the contact details you provided." And finally we have the closing P tag,.

File: app/views/enquiry_submitted.scala.html [hunk 8]
```diff
@@ -27,13 +29,17 @@
...
     <h2 class="heading-medium">@messages("enquiry.success.next.heading")</h2>
-    @components.text("enquiry.success.next.para")
-    @components.text("enquiry.success.next.para1")
-    @components.text("enquiry.success.next.para2")
-    <p>@messages("enquiry.success.next.para3") <a href="@routes.EnquiryController.messagesRedirect()">
-       @messages("enquiry.success.next.link_text")</a> @messages("enquiry.success.next.para4")</p>
-    <a id="start-again" href="@routes.EnquiryController.personalAccountRedirect()" class="button">@messages("site.account.home")</a>
+    <p>@components.text("enquiry.success.next.para1") @responseTime @components.text("enquiry.success.next.para1_1")</p>
+    <p>@components.text("enquiry.success.next.para2")</p>
+    <p>@{
```
Comment (rwalpole) [inline app/views/enquiry_submitted.scala.html:36]: Extra `<p>` tag.

File: app/views/enquiry_submitted.scala.html [hunk 9]
```diff
@@ -27,13 +29,17 @@
...
+    @components.text("enquiry.success.note.para1")
     <h2 class="heading-medium">@messages("enquiry.success.next.heading")</h2>
-    @components.text("enquiry.success.next.para")
-    @components.text("enquiry.success.next.para1")
-    @components.text("enquiry.success.next.para2")
-    <p>@messages("enquiry.success.next.para3") <a href="@routes.EnquiryController.messagesRedirect()">
-       @messages("enquiry.success.next.link_text")</a> @messages("enquiry.success.next.para4")</p>
-    <a id="start-again" href="@routes.EnquiryController.personalAccountRedirect()" class="button">@messages("site.account.home")</a>
+    <p>@components.text("enquiry.success.next.para1") @responseTime @components.text("enquiry.success.next.para1_1")</p>
+    <p>@components.text("enquiry.success.next.para2")</p>
```
Comment (rwalpole) [inline app/views/enquiry_submitted.scala.html:35]: Extra `<p>` tags.

Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:35]: Yes. That's correct.

File: app/views/enquiry_submitted.scala.html (hunk 8 above)
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:36]: Yes correct.

Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:36]: The old version was a button. This version is a pure anchor link.

File: app/views/enquiry_submitted.scala.html (hunk 9 above)
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:35]: Let me send a screenshot to Michael, Chucks and Jim and if they need anything changed I will amend the PR.

File: app/views/enquiry_submitted.scala.html (hunk 8 above)
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:36]: I'll send a screengrab to Michael and the others and see if they're happy.

Comment (elabeca) [reply app/views/enquiry_submitted.scala.html]: @rwalpole I removed the P tag around the bottom return url as you suggested as the extra spacing was marginal.

File: app/views/enquiry_submitted.scala.html (hunk 9 above)
Comment (elabeca) [reply app/views/enquiry_submitted.scala.html:35]: Michael and Chuks are happy with the said output. They've raised other concerns around existing copy but that's now been fixed.

File: app/controllers/EnquiryController.scala [hunk 10]
```diff
@@ -61,23 +63,30 @@ class EnquiryController @Inject()(
...
             (errorPage: Result) => errorPage,
-            details =>
-              Ok(
-                enquiry(
-                  appConfig,
-                  form,
-                  EnquiryDetails(enquiryType, "", "", email, "", details.taxId),
-                  details.responseTime
+            details => {
+              validateEncryptedReturnLink(returnLink).fold(
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:67]: this line and line 64 you have two folds on either, wouldn't a for comprehension make the code more readable?

File: app/controllers/EnquiryController.scala [hunk 11]
```diff
@@ -117,7 +128,16 @@ class EnquiryController @Inject()(
           case CREATED =>
             extractId(response) match {
               case Right(id) =>
-                Ok(enquiry_submitted(appConfig, id.id, responseTime))
+                validateEncryptedReturnLink(returnLink).fold(
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:131]: same thing as above, can't you use for comprehension to for this Either and the extractId one ?

File: app/controllers/EnquiryController.scala [hunk 12]
```diff
@@ -139,14 +159,18 @@ class EnquiryController @Inject()(
...
-
-  def personalAccountRedirect: Action[AnyContent] =
-    Action {
-      Redirect(appConfig.personalAccountUrl)
+  private def validateEncryptedReturnLink(returnLink: Option[ReturnLink]): Either[Result, Option[ReturnLink]] =
+    try {
+      Right(
+        returnLink.map(link => {
+          ReturnLink(
+            crypto.QueryParameterCrypto.decrypt(Crypted(link.returnUrl)).value,
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:167]: do you validate this is a valid Url ?

File: app/controllers/EnquiryController.scala [hunk 13]
```diff
@@ -139,14 +159,18 @@ class EnquiryController @Inject()(
...
+            crypto.QueryParameterCrypto.decrypt(Crypted(link.returnLinkText)).value
```
Comment (OlegEfrem) [inline app/controllers/EnquiryController.scala:168]: how do you know this is not an empty string?

File: app/controllers/binders/Binders.scala [hunk 14]
```diff
@@ -66,4 +66,29 @@ object Binders {
...
+      ): Option[Either[String, ReturnLink]] = {
+
+        val returnUrl = stringBinder.bind("returnUrl", params)
+        val returnLinkText = stringBinder.bind("returnLinkText", params)
+
+        (returnUrl, returnLinkText) match {
+          case (Some(Right(returnUrl)), Some(Right(returnLinkText))) =>
+            Some(Right(ReturnLink(returnUrl, returnLinkText)))
+          case (None, None) => None
+          case _            => Some(Left("No return link provided"))
```
Comment (OlegEfrem) [inline app/controllers/binders/Binders.scala:86]: I think here you would want to propagate the Left from bind on line 79/80, isn't it?
as such, I think a for comprehension or maybe even an EitherT from cats would be way more readable.

File: app/controllers/binders/Binders.scala [hunk 15]
```diff
@@ -66,4 +66,29 @@ object Binders {
...
+        }
+      }
+
+      override def unbind(key: String, returnLink: ReturnLink): String =
+        stringBinder.unbind("returnUrl", returnLink.returnUrl) + "&" + stringBinder
+          .unbind("returnLinkText", returnLink.returnLinkText)
```
Comment (OlegEfrem) [inline app/controllers/binders/Binders.scala:92]: think we need to be consistent either we name both parameters returnUrl & returnUrlText, or name them returnLinkUrl & returnLinkText, i would go for the latter to match the suggestion for the case class.

File: app/views/govuk_wrapper.scala.html [hunk 16]
```diff
@@ -41,7 +41,7 @@
 
 @insideHeader = {
     @uiLayouts.header_nav(
-      navTitle = Some(messages("site.service_name")),
+      navTitle = Some(messages("site.service.name.bta")),
```
Comment (OlegEfrem) [inline app/views/govuk_wrapper.scala.html:44]: did we say we want to keep P800 in the PTA space?

File: conf/application.conf [hunk 17]
```diff
@@ -169,9 +169,4 @@ contact-frontend {
 mongodb {
   uri = "mongodb://localhost:27017/"${appName}
   timeToLiveInSeconds = 3600
-}
-
-urls {
```
Comment (OlegEfrem) [inline conf/application.conf:174]: why did we remove them?

File: conf/app.routes [hunk 18]
```diff
@@ -4,10 +4,8 @@ GET     /assets/*file                                   controllers.Assets.versi
 
 GET     /language/:lang                                 controllers.LanguageSwitchController.switchToLanguage(lang: String)
 
-GET     /message/$enquiryType<sa-general|ct-general|vat-general|epaye-general|epaye-jrs>/make_enquiry      controllers.EnquiryController.onPageLoad(enquiryType: String)
-POST    /message/submit                                 controllers.EnquiryController.onSubmit()
-GET     /messages                                       controllers.EnquiryController.messagesRedirect()
-GET     /messages/personal-account                      controllers.EnquiryController.personalAccountRedirect()
```
Comment (OlegEfrem) [inline conf/app.routes:10]: we did we remove line 9 & 10 ?

File: it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala [hunk 19]
```diff
@@ -147,18 +148,197 @@ class EnquiryControllerFrontendSpec
...
+      emailField("email").value = "test@dummy.com"
+      textArea("question").value = "A question from the customer"
+      click on find(id("submit")).value
+
+      eventually {
+        pageSource must include(
+          "<a id=\"redirect-return-link\" href=\"foo&amp;value\" target=\"_self\" data-sso=\"false\">")
+        pageSource must include("bar")
+      }
+    }
```
Comment (OlegEfrem) [inline it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala:217]: can we extract repeated code into re-usable functions and leave in the test scenarios only the relevant things, ideally no more than a couple of lines of code?

File: test/controllers/binders/BindersSpec.scala [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    "fail if the returnUrl is not present" in {
+      controllers.binders.Binders.queryStringReturnLinkBinder.bind("anyValName", Map(validReturnLinkText)) should
+        be(Some(Left("No return link provided")))
+    }
+
+    "fail if the returnLinkText is not present" in {
+      controllers.binders.Binders.queryStringReturnLinkBinder.bind("anyValName", Map(validReturnUrl)) should
+        be(Some(Left("No return link provided")))
```
Comment (OlegEfrem) [inline test/controllers/binders/BindersSpec.scala:44]: can you add a test if it is an empty string?

File: app/models/ReturnLink.scala [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+case class ReturnLink(returnUrl: String, returnLinkText: String)
```
[the remark this answers is not in the document]
Comment (elabeca) [reply app/models/ReturnLink.scala:19]: I agree. Good point. I was mirroring the parameter names to preferences-frontend and also wanted to have a direct query parameter relationship (as these are query params keys).

File: app/controllers/EnquiryController.scala (hunk 10 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala:67]: I was wondering about that earlier but pure Either's aren't Monads and I wanted to do something with each of their Left results (in this case, two very distinct error results). I'll have a look if I can make it slighly more readible.

File: app/controllers/EnquiryController.scala (hunk 12 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala:167]: no I don't. That method should be called decrypt rather than validate

File: app/controllers/EnquiryController.scala (hunk 13 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala:168]: this method is meant to return a BadRequest if the decryption fails. Both of these are paramters are validated in controllers.Binders.queryStringReturnLinkBinder()

File: app/models/ReturnLink.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (OlegEfrem) [reply app/models/ReturnLink.scala:19]: If we have to choose between consistency across frontend and getting it right, I'd favor the latter :)

File: app/views/govuk_wrapper.scala.html (hunk 16 above)
Comment (elabeca) [reply app/views/govuk_wrapper.scala.html:44]: A quick discussion with Michael, Jim and Chuks this morning, we agreed that making this title dynamic to handle both bta and pta could be a seperate story not to extend this current story with scope creep further than it already is.

File: conf/application.conf (hunk 17 above)
Comment (elabeca) [reply conf/application.conf:174]: these are no longer used throughout the application as we're now offering dynamic return urls.

File: app/controllers/EnquiryController.scala (hunk 10 above)
Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:67]: from what I see the error type is the same for both: Result, did I miss something?

File: conf/app.routes (hunk 18 above)
Comment (elabeca) [reply conf/app.routes:10]: no longer required. these endpoints are only used internally by the application. /messages/personal-account did a redirect to bta-personal account which isn't required anymore with the new logic in the view.

File: app/controllers/EnquiryController.scala (hunk 12 above)
Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:167]: ok, what happens if we get an invalid URL or they just mix up the text with the URL?

File: it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala (hunk 19 above)
Comment (elabeca) [reply it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala:217]: totally agree.

File: app/views/govuk_wrapper.scala.html (hunk 16 above)
Comment (OlegEfrem) [reply app/views/govuk_wrapper.scala.html:44]: i mean previously this would go to PTA, now we go to BTA, I thought we wanted to keep previous functionality intact or I missed something?

File: app/controllers/EnquiryController.scala (hunk 13 above)
Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:168]: ok, you have a comment there then :)

File: app/controllers/binders/Binders.scala (hunk 14 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala:86]: Doing the non empty validation this instant. And I'm glad you've brought the subject of cats I'm sure @lxol will be pleased

File: app/controllers/EnquiryController.scala (hunk 12 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala:167]: do we have a requirement to validate urls provided by our internal services? if we do, I'm more than happy to implement something around that.

File: app/controllers/binders/Binders.scala (hunk 15 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala:92]: I was only mirroring our other service preferences-frontend.

File: app/controllers/EnquiryController.scala (hunk 12 above)
Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:167]: well, these parameters would be provided by external clients, and I'd say we need zero trust principle here, validate everything that is possible to avoid problems later. 
I will add this as a requirement.

File: test/controllers/binders/BindersSpec.scala (hunk 20 above)
Comment (OlegEfrem) [reply test/controllers/binders/BindersSpec.scala:44]: also if it is an invalid URL.

File: app/controllers/EnquiryController.scala (hunk 11 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala:131]: not in this case as Either's don't have flatMap unless you make them right biased, but I'll have a look, I need to do something very specific with two different Leftie return types, it's not just a matter of return a generic error message. I'll maybe transform the signature validateEncryptedReturnLink so that it matches the left return type of exactId.

File: test/controllers/binders/BindersSpec.scala (hunk 20 above)
Comment (elabeca) [reply test/controllers/binders/BindersSpec.scala]: done.

File: app/views/govuk_wrapper.scala.html (hunk 16 above)
Comment (elabeca) [reply app/views/govuk_wrapper.scala.html:44]: Yes, that's correct, albeit p800 is disabled, we wanted to simply be able to re-enable the routes and it would work seemlessly. However, it's not an easy change. So it was decided to create a seperate story for it. Basically we need to feed the enquiry type to the gouk wrapper. Somehow.

File: app/controllers/EnquiryController.scala (hunk 11 above)
Comment (OlegEfrem) [reply app/controllers/EnquiryController.scala:131]: yes, this is what i was thinking of, to make extractId Left return Response by moving it's Left logic there.

File: app/views/govuk_wrapper.scala.html (hunk 16 above)
Comment (OlegEfrem) [reply app/views/govuk_wrapper.scala.html:44]: makes sense, marking this comment as resolved.

File: test/controllers/binders/BindersSpec.scala (hunk 20 above)
Comment (OlegEfrem) [reply test/controllers/binders/BindersSpec.scala]: forgot to ask about invalid URL test :) sorry :)

File: app/controllers/binders/Binders.scala (hunk 14 above)
Comment (OlegEfrem) [reply app/controllers/binders/Binders.scala:91]: i think you can you or matching to reduce the number of cases: 
case (Some(Right(_)), Some(Right(""))) **|** (Some(Right(_)), None) => Some(Left("Missing return link text"))

File: test/controllers/binders/BindersSpec.scala (hunk 20 above)
Comment (elabeca) [reply test/controllers/binders/BindersSpec.scala]: I've added this test in EnquiryControllerFrontendSpec.scala as the URI validation happens after the binding (when it gets decrypted at the very last moment that we need it decrypted)

File: app/controllers/EnquiryController.scala (hunk 12 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala]: Done. added URI validation to returnUrl.

File: app/models/ReturnLink.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (elabeca) [reply app/models/ReturnLink.scala]: Done. New signature/param names are: ReturnLinkt(url: String, text: String)

File: app/controllers/EnquiryController.scala (hunk 10 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala]: Done.

File: app/controllers/binders/Binders.scala (hunk 14 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala:91]: @OlegEfrem I'm about to refactor this section of code:

```
          case (Some(Right("")), Some(Right(""))) => Some(Left("No return link provided"))
          case (Some(Right(_)), Some(Right(""))) => Some(Left("Missing return link text"))
          case (Some(Right("")), Some(Right(_))) => Some(Left("Missing return url"))
          case (Some(Right(_)), None) => Some(Left("Missing return link text"))
          case (None, Some(Right(_))) => Some(Left("Missing return url"))
          case (Some(Right(returnUrl)), Some(Right(returnLinkText))) =>
            Some(Right(ReturnLink(returnUrl, returnLinkText)))
          case (None, None) => None
          case _            => Some(Left("No return link provided"))
```

could we describe on the story what are the error messages should be (if there should be many different ones, or just the one generic) and update the acceptance criterias so that our testers can test? thanks.

File: app/controllers/EnquiryController.scala (hunk 11 above)
Comment (elabeca) [reply app/controllers/EnquiryController.scala]: Done.

File: app/controllers/binders/Binders.scala (hunk 14 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala:91]: @OlegEfrem I've updated the story acceptance criterias around return link error handling in scenario 7-12. This could be captured into a table/matrix, but for now it's just high level. If you could review and make amendments to the story so I can reflect those changes in my PR. Thanks

File: it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala (hunk 19 above)
Comment (elabeca) [reply it/uk/gov/hmrc/twowaymessagefrontend/EnquiryControllerFrontendSpec.scala:225]: - [ ] refactor tests to use common scenario builder

File: app/controllers/binders/Binders.scala (hunk 14 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala:91]: - [ ] update with new messages and refactor if possible

File: app/controllers/binders/Binders.scala (hunk 15 above)
Comment (elabeca) [reply app/controllers/binders/Binders.scala]: - [ ] rename query parameters to returnLinkUrl & returnLinkText

**Comment by elabeca** [comment]
- [ ] update readme
